client/transaction-pool/src/lib.rs

```diff
@@ -54,6 +60,55 @@ pub struct BasicPool<PoolApi, Block>
 	api: Arc<PoolApi>,
 	revalidation_strategy: Arc<Mutex<RevalidationStrategy<NumberFor<Block>>>>,
 	revalidation_queue: Arc<revalidation::RevalidationQueue<PoolApi>>,
+	ready_poll: Arc<Mutex<ReadyPoll<ReadyIteratorFor<PoolApi>, Block>>>,
+}
+
+struct ReadyPoll<T, Block>
+where
+	Block: BlockT,
+{
+	updated_at: Option<NumberFor<Block>>,
+	pollers: Vec<(NumberFor<Block>, futures::channel::oneshot::Sender<T>)>,
+}
+
+impl<T, Block> Default for ReadyPoll<T, Block>
+where
+	Block: BlockT,
+{
+	fn default() -> Self {
+		Self {
+			updated_at: None,
+			pollers: Default::default(),
+		}
+	}
+}
+
+impl<T, Block> ReadyPoll<T, Block>
+where
+	Block: BlockT,
+{
+	fn trigger(&mut self, number: NumberFor<Block>, iterator_factory: impl Fn() -> T) {
+		self.updated_at = Some(number);
+		let pollers = self.pollers.drain(..).collect::<Vec<(_, _)>>();
+
+		for (poller_number, poller_sender) in pollers {
```

Comment:
That loop look like it could just be `self.pollers.retain()` no?

Reply:
oneshot `Sender::send` is by-value